M365/MDO/ResendFailedMail.ps1

```diff
@@ -0,0 +1,322 @@
+                $recipientCache[$failedMessage.SenderAddress] = $false
+            } else {
+                Write-Verbose "Added to cache Recipient $($failedMessage.SenderAddress) with Id $($failedMessage.SenderAddress)"
+                $recipientCache[$failedMessage.SenderAddress] = $true
+                Write-Verbose "Verified $($failedMessage.SenderAddress)"
+                $verifiedAcceptedSenderMessages += $failedMessage
+            }
+        } catch {
+            Write-Verbose "Error getting Sender Address $($failedMessage.SenderAddress)"
+            Write-Verbose "Discarded $($failedMessage.MessageId) - Subject: $($failedMessage.Subject)"
+            $recipientCache[$failedMessage.SenderAddress] = $false
+            Write-Verbose "$_"
+        }
+    }
+}
+
+$totalMessages = $verifiedAcceptedSenderMessages.Count
+if ($totalMessages -gt 0) {
+    if (-not $Force) {
+        Write-Host "`nWe are going to resend the following messages:"
+        Write-Host ($verifiedAcceptedSenderMessages | Format-Table -AutoSize Received, MessageId, SenderAddress, RecipientAddress, Subject | Out-String)
+        Write-Host "Total number of messages: $totalMessages`n"
+    }
+
+    if ($Force -or $PSCmdlet.ShouldContinue("Are you sure you want to do it?", "Resend messages")) {
+        $count = 0
+        $resendCount = 0
+        foreach ( $failedMessage in $verifiedAcceptedSenderMessages ) {
+            $count++
+            Write-Progress -Activity "Resending Progress" -Status "$count of $totalMessages" -PercentComplete ($count / $totalMessages * 100) -CurrentOperation "Resending message $($failedMessage.MessageId) - Subject: $($failedMessage.Subject)"
+            try {
+                $fullMessage = $null
+                $fullMessage = Get-MgUserMessage -UserId $failedMessage.SenderAddress -Filter "InternetMessageId eq '$($failedMessage.MessageId)'" -ExpandProperty Attachments -ErrorAction Stop | Sort-Object ReceivedDateTime | Select-Object -First 1
+            } catch {
+                Write-Host "Error getting message $($failedMessage.MessageId) - Subject: $($failedMessage.Subject)" -ForegroundColor Red
+                Write-Verbose "$_"
+            }
+            if ($fullMessage.Count -eq 0) {
+                Write-Host "Message not found for $($failedMessage.MessageId)" -ForegroundColor Yellow
+            } else {
+                Write-Verbose "Resending message $($failedMessage.MessageId) - Subject: $($fullMessage.Subject)"
+                try {
+                    Send-MgUserMessage -UserId $failedMessage.SenderAddress -MessageId $fullMessage.Id -ErrorAction Stop
+                    Write-Host "Resent Message: $($failedMessage.MessageId) - Subject: $($fullMessage.Subject)"
+                    $resendCount++
+                } catch {
+                    Write-Host "Error resending message $($failedMessage.MessageId) - Subject: $($fullMessage.Subject)" -ForegroundColor Red
+                    Write-Verbose "$_"
```

Comment:
Because we are already at the end of the loop, you don't need to have a `continue` here.